datasetinsights/stats/visualization/bbox3d_plot.py

```diff
@@ -0,0 +1,146 @@
+
+    _cv2.line(image, ful, bul, color, box_line_width)  # top left
+    _cv2.line(image, fur, bur, color, box_line_width)  # top right
+    _cv2.line(image, fbl, bbl, color, box_line_width)  # bottom left
+    _cv2.line(image, fbr, bbr, color, box_line_width)  # bottom right
+
+
+def add_single_bbox3d_on_image(
+        image,
+        box,
+        projection,
+        color=None,
+        box_line_width=2,
+):
+    """" Add single 3D bounding box on a given image.
+
+    Args:
+        image (numpy array): a numpy array for an image
+        box (BBox3D): a 3D bounding box in camera's coordinate system
+        projection (numpy 2D array): camera's 3x3 projection matrix
+        color(tuple): RGBA color of the bounding box. Defaults to None. If color = None
+        the the tuple of [0, 255, 0, 255] (Green) will be used.
+        box_line_width (int): line width of the bounding boxes. Defaults to 2.
+    """
+    img_height, img_width, _ = image.shape
+
+    fll = box.back_left_bottom_pt
+    ful = box.back_left_top_pt
+    fur = box.back_right_top_pt
+    flr = box.back_right_bottom_pt
+
+    bll = box.front_left_bottom_pt
+    bul = box.front_left_top_pt
+    bur = box.front_right_top_pt
+    blr = box.front_right_bottom_pt
+
+    fll_raster = _project_pt_to_pixel_location(fll, projection, img_height, img_width)
+    ful_raster = _project_pt_to_pixel_location(ful, projection, img_height, img_width)
+    fur_raster = _project_pt_to_pixel_location(fur, projection, img_height, img_width)
+    flr_raster = _project_pt_to_pixel_location(flr, projection, img_height, img_width)
+    bll_raster = _project_pt_to_pixel_location(bll, projection, img_height, img_width)
+    bul_raster = _project_pt_to_pixel_location(bul, projection, img_height, img_width)
+    bur_raster = _project_pt_to_pixel_location(bur, projection, img_height, img_width)
+    blr_raster = _project_pt_to_pixel_location(blr, projection, img_height, img_width)
+
+    _add_single_bbox3d_on_image(
+        image,
+        fll_raster,
+        ful_raster,
+        fur_raster,
+        flr_raster,
+        bll_raster,
+        bul_raster,
+        bur_raster,
+        blr_raster,
+        color,
+        box_line_width
+    )
+
+def _project_pt_to_pixel_location(pt, projection, img_height, img_width):
+    """ Projects a 3D coordinate into a pixel location.
+
+    Applies the passed in projection matrix to project a point from the camera's
+    coordinate space into pixel space.
+
+    Args:
+        pt (numpy array): The 3D point to project.
+        projection (numpy 2D array): The camera's 3x3 projection matrix.
+        img_height (int): The height of the image in pixels.
+        img_width (int): The width of the image in pixels.
+
+    Returns:
+        numpy array: a one-dimensional array with two values (x and y) representing a point's
+        pixel coordinate in an image.
+    """
+
+    _pt = projection.dot(pt)
+    _pt /= _pt[2]
```

Comment:
What is this divison for, out of curiosity? Isn't it possible for the projected point to be at depth 0?

Reply:
It's the formula for perspective divide which is used to map cartesian coordinates to homogeneous/projected coordinates, see: https://www.scratchapixel.com/lessons/3d-basic-rendering/computing-pixel-coordinates-of-3d-point/mathematics-computing-2d-coordinates-of-3d-points.

The theory is, from my reading, that divide by 0 should always be prevented by the near clipping plane, but it is a good catch and we should protect against that case.

Reply:
Got it! I learned something today.